src/coreclr/jit/codegenloongarch64.cpp

```diff
@@ -7660,46 +7531,50 @@ void CodeGen::instGen_MemoryBarrier(BarrierKind barrierKind)
  *      ...
  *      st.d s8,sp,off2+8*8
  *
- *      st.d ra,sp,off3
- *      st.d fp,sp,off3+8
+ *      st.d ra,sp,off3+8
+ *      st.d fp,sp,off3
  *
  * Notes:
  * 1. FP is always saved, and the first store is FP, RA.
  * 2. General-purpose registers are 8 bytes, floating-point registers are 8 bytes.
  * 3. For frames with varargs, not implemented completely and not tested !
  * 4. We allocate the frame here; no further changes to SP are allowed (except in the body, for localloc).
  *
- * For functions with GS and localloc, we change the frame so the frame pointer and RA are saved at the top
- * of the frame, just under the varargs registers (if any). Note that the funclet frames must follow the same
- * rule, and both main frame and funclet frames (if any) must put PSPSym in the same offset from Caller-SP.
+ * For functions with GS and localloc, we had saved the frame pointer and RA at the top
+ * of the frame. Note that the funclet frames must follow the same rule,
+ * and both main frame and funclet frames (if any) must put PSPSym in the same offset from Caller-SP.
  * Since this frame type is relatively rare, we force using it via stress modes, for additional coverage.
  *
  * The frames look like the following (simplified to only include components that matter for establishing the
  * frames). See also Compiler::lvaAssignFrameOffsets().
  *
- *
  * The LoongArch64's frame layout is liking:
  *
+ *    If we need to generate a GS cookie, we need to make sure the saved frame pointer and return address
+ *    (FP and RA) are protected from buffer overrun by the GS cookie.
+ *    So we always save the FP/RA along with the rest of the callee-saved registers above.
+ *
  *      |                       |
  *      |-----------------------|
  *      |  incoming arguments   |
  *      +=======================+ <---- Caller's SP
- *      |     Arguments  Or     | // if needed.
  *      |  Varargs regs space   | // Only for varargs functions; (varargs not implemented for LoongArch64)
  *      |-----------------------|
  *      |    MonitorAcquired    | // 8 bytes; for synchronized methods
  *      |-----------------------|
- *      |        PSP slot       | // 8 bytes (omitted in NativeAOT ABI)
+ *      |        PSPSym         | // 8 bytes, Only for frames with EH, (omitted in NativeAOT ABI)
  *      |-----------------------|
- *      | locals, temps, etc.   |
+ *      |Callee saved registers | // not including FP/RA; multiple of 8 bytes
  *      |-----------------------|
- *      |  possible GS cookie   |
+ *      |      Saved RA         | // 8 bytes
  *      |-----------------------|
  *      |      Saved FP         | // 8 bytes
```

Comment:
@jakobbotsch 
I just move the FP/RA to the top frame which always saved with calleeSavedRegs together.

Reply:
While the `MonitorAcquired` and `PSPSym` on the top of the frame.  The layout liking this PR's first comment description.

Reply:
"possible GS cookie" appears twice in this table. Should one be removed?

Reply:
> "possible GS cookie" appears twice in this table. Should one be removed?

The ARMARCH/XARCH also like this as the different condition to save it at different position within the lclvar.cpp.

Reply:
> > "possible GS cookie" appears twice in this table. Should one be removed?
> 
> The ARMARCH/XARCH also like this as the different condition to save it at different position within the lclvar.cpp.

The line 5488 within lclvars.cpp.